README.md
# wit-ruby

`wit-ruby` is the Ruby SDK for [Wit.ai](http://wit.ai).

## Install

From RubyGems:
```bash
gem install wit
```

From source:
```bash
git clone https://github.com/wit-ai/wit-ruby
gem build wit.gemspec
gem install wit-*.gem
```

## Quickstart

Run in your terminal:

```bash
ruby examples/quickstart.rb <your_token>
```

See the `examples` folder for more examples.

## API

### Overview

`wit-ruby` provides a Wit class with the following methods:
* `message` - the Wit [message API](https://wit.ai/docs/http/20160330#get-intent-via-text-link)
* `interactive` - starts an interactive conversation with your bot

### Wit class

The Wit constructor takes a `Hash` with the following symbol keys:
* `:access_token` - the access token of your Wit instance
* `:timeout` - the timeout duration in seconds. the class will raise a Net::ReadTimeout when the duration has expired.

A minimal example looks like this:
```ruby
require 'wit'

client = Wit.new(access_token: access_token)
client.message('set an alarm tomorrow at 7am')
```

### .message()

The Wit [message API](https://wit.ai/docs/http/20160330#get-intent-via-text-link).

Takes the following parameters:
* `msg` - the text you want Wit.ai to extract the information from

Example:
```ruby
rsp = client.message('what is the weather in London?')
puts("Yay, got Wit.ai response: #{rsp}")
```

### .interactive()

Starts an interactive conversation with your bot.

Example:
```ruby
client.interactive
```

### .run_actions()

**DEPRECATED** See [our blog post](https://wit.ai/blog/2017/07/27/sunsetting-stories) for a migration plan.

A higher-level method to the Wit converse API.
`run_actions` resets the last turn on new messages and errors.

Takes the following parameters:
* `session_id` - a unique identifier describing the user session
* `message` - the text received from the user
* `context` - the `Hash` representing the session state
* `max_steps` - (optional) the maximum number of actions to execute (defaults to 5)

Example:
```ruby
session = 'my-user-session-42'
context0 = {}
context1 = client.run_actions(session, 'what is the weather in London?', context0)
p "The session state is now: #{context1}"
context2 = client.run_actions(session, 'and in Brussels?', context1)
p "The session state is now: #{context2}"
```

### .converse()

**DEPRECATED** See [our blog post](https://wit.ai/blog/2017/07/27/sunsetting-stories) for a migration plan.

The low-level Wit [converse API](https://wit.ai/docs/http/20160330#converse-link).

Takes the following parameters:
* `session_id` - a unique identifier describing the user session
* `msg` - the text received from the user
* `context` - the `Hash` representing the session state
* `reset` - (optional) whether to reset the last turn

Example:
```ruby
rsp = client.converse('my-user-session-42', 'what is the weather in London?', {})
puts("Yay, got Wit.ai response: #{rsp}")
```

### CRUD operations for entities
payload in the parameters is a hash containing API arguments

#### .get_entities()
Returns a list of available entities for the app.  
See [GET /entities](https://wit.ai/docs/http/20160526#get--entities-link)

#### .post_entities(payload)
Creates a new entity with the given attributes.  
See [POST /entities](https://wit.ai/docs/http/20160526#post--entities-link)

#### .get_entity(entity_id)
Returns all the expressions validated for an entity.  
See [GET /entities/:entity-id](https://wit.ai/docs/http/20160526#get--entities-:entity-id-link)

#### .put_entities(entity_id, payload)
Updates an entity with the given attributes.  
See [PUT /entities/:entity-id](https://wit.ai/docs/http/20160526#put--entities-:entity-id-link)

#### .delete_entities(entity_id)
Permanently remove the entity.  
See [DELETE /entities/:entity-id](https://wit.ai/docs/http/20160526#delete--entities-:entity-id-link)

#### .post_values(entity_id, payload)
Add a possible value into the list of values for the entity.  
See [POST /entities/:entity-id/values](https://wit.ai/docs/http/20160526#post--entities-:entity-id-values-link)

#### .delete_values(entity_id, value)
Delete a canonical value from the entity.  
See [DELETE /entities/:entity-id/values/:value](https://wit.ai/docs/http/20160526#delete--entities-:entity-id-values-link)

#### post_expressions(entity_id, value, payload)
Create a new expression of the canonical value of the entity.  
See [POST /entities/:entity-id/values/:value/expressions](https://wit.ai/docs/http/20160526#post--entities-:entity-id-values-:value-id-expressions-link)

#### delete_expressions(entity_id, value, expression)
Delete an expression of the canonical value of the entity.  
See [DELETE /entities/:entity-id/values/:value/expressions/:expression](https://wit.ai/docs/http/20160526#delete--entities-:entity-id-values-:value-id-expressions-link)

See the [docs](https://wit.ai/docs) for more information.

### Logging

Default logging is to `STDOUT` with `INFO` level.

You can setup your logging level as follows:
```ruby
Wit.logger.level = Logger::WARN
```
See the [Logger class](http://ruby-doc.org/stdlib-2.1.0/libdoc/logger/rdoc/Logger.html) docs for more information.

## Thanks

Thanks to [Justin Workman](http://github.com/xtagon) for releasing a first version in October 2013. We really appreciate!


CHANGES.md
## v5.1.0

- adding timeout as a contructor option

## v5.0.0

- `converse` and `run_actions` are deprecated
- `interactive` now calls `message`

### Breaking changes
- Renamed WitException to Wit::Error
- Changed Wit::Error to inherit from StandardError instead of Exception
- Moved constants inside Wit namespace
- Moved #req and #validate_actions to private methods within Wit namespace

## v4.1.0

- `converse` now takes `reset` as optional parameter.

### Breaking changes

- `run_actions` now resets the last turn on new messages and errors.

## v4.0.0

After a lot of internal dogfooding and bot building, we decided to change the API in a backwards-incompatible way. The changes are described below and aim to simplify user code and accommodate upcoming features.

See `./examples` to see how to use the new API.

### Breaking changes

- `say` renamed to `send` to reflect that it deals with more than just text
- Removed built-in actions `merge` and `error`
- Actions signature simplified with `request` and `response` arguments
- INFO level replaces LOG level

## v3.4

- allows for overriding API version, by setting `WIT_API_VERSION`
- `interactive()` mode
- warns instead of throwing when validating actions

### breaking

- bumped default API version from `20160330` to `20160516`

## v3.3.1

- fixed recursive call when running actions
- throws an exception if context passed is not a Hash

## v3.3

Unifying action parameters

### breaking

- the `say` action now takes 3 parameters: `session_id`, `context`, `msg`
- the `error` action now takes 3 parameters: `session_id`, `context`, `error`

## v3.2

Updating action parameters

### breaking

- the `merge` action now takes 4 parameters: `session_id`, `context`, `entities`, `msg`
- the `error` action now takes `context` as second parameter
- custom actions now take 2 parameters: `session_id`, `context`

## v3.1

- allows for custom logging
- better error message

## v3.0

Bot Engine integration

### breaking

- the `message` API is wrapped around a `Wit` class, and doesn't take the token as first parameter

## v2.0

Rewrite in pure Ruby

### breaking

- audio recording and streaming have been removed because:
  - many people only needed access to the HTTP API, and audio recording did not make sense for server-side use cases
  - dependent on platform, choice best left to developers
  - forced us to maintain native bindings as opposed to a pure Pythonic library
- we renamed the functions to match the HTTP API more closely
  - `.text_query(string, access_token)` becomes `.message(access_token, string)`
- all functions now return a Ruby dict instead of a JSON string


spec/wit_spec.rb
require 'wit'
RSpec.describe Wit do
  context "req" do
    let(:wit) { Wit.new(timeout: 3, access_token: 'some_token') }

    it "raises Net::ReadTimeout exception" do
      stub_request(:any, /https:\/\/api\.wit\.ai*+/)
      .to_timeout

      expect do
        wit.message("some text message")
      end.to raise_error(Net::OpenTimeout)
    end

    it "returns expected json" do
      stub_request(:any, /https:\/\/api\.wit\.ai*+/)
      .to_return(status: 200, body: {}.to_s, headers: {})

      expect(wit.message("some text message")).to eq({})
    end
  end
end


spec/spec_helper.rb
require 'webmock/rspec'
WebMock.disable_net_connect!

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


lib/wit.rb
require 'json'
require 'logger'
require 'net/http'
require 'securerandom'

class Wit
  class Error < StandardError; end

  WIT_API_HOST = ENV['WIT_URL'] || 'https://api.wit.ai'
  WIT_API_VERSION = ENV['WIT_API_VERSION']  || '20160516'
  DEFAULT_MAX_STEPS = 5
  LEARN_MORE = 'Learn more at https://wit.ai/docs/quickstart'
  DEFAULT_TIMEOUT = 60

  def initialize(opts = {})
    @access_token = opts[:access_token]

    if opts[:logger]
      @logger = opts[:logger]
    end

    if opts[:actions]
      logger.warn('Stories and POST /converse have been deprecated. This will break in February 2018!')
      @actions = validate_actions(logger, opts[:actions])
    end
    @timeout = opts[:timeout] || DEFAULT_TIMEOUT
    @_sessions = {}
  end

  def logger
    @logger ||= begin
      x = Logger.new(STDOUT)
      x.level = Logger::INFO
      x
    end
  end

  def message(msg, n=nil, verbose=nil)
    params = {}
    params[:q] = msg unless msg.nil?
    params[:n] = n unless n.nil?
    params[:verbose] = verbose unless verbose.nil?
    res = req(logger, @access_token, Net::HTTP::Get, '/message', params)
    return res
  end

  def converse(session_id, msg, context={}, reset=nil)
    if !context.is_a?(Hash)
      raise Error.new('context should be a Hash')
    end
    params = {}
    params[:q] = msg unless msg.nil?
    params[:session_id] = session_id
    params[:reset] = true if reset
    res = req(logger, @access_token, Net::HTTP::Post, '/converse', params, context)
    return res
  end

  def run_actions(session_id, message, context={}, max_steps=DEFAULT_MAX_STEPS)
    if !@actions
      throw_must_have_actions
    end
    if !context.is_a?(Hash)
      raise Error.new('context should be a Hash')
    end

    # Figuring out whether we need to reset the last turn.
    # Each new call increments an index for the session.
    # We only care about the last call to run_actions.
    # All the previous ones are discarded (preemptive exit).
    current_request = 1
    if @_sessions.has_key?(session_id)
      current_request = @_sessions[session_id] + 1
    end
    @_sessions[session_id] = current_request

    context = __run_actions(session_id, current_request, message, context, max_steps)

    # Cleaning up once the last call to run_actions finishes.
    if current_request == @_sessions[session_id]
      @_sessions.delete(session_id)
    end

    return context
  end

  def interactive(context={})
    while true
      print '> '
      msg = STDIN.gets.strip
      if msg == ''
        next
      end

      begin
        puts(message(msg))
      rescue Error => exp
        logger.error("error: #{exp.message}")
      end
    end
  rescue Interrupt => _exp
    puts
  end

  def throw_if_action_missing(action_name)
    if !@actions.has_key?(action_name)
      raise Error.new("unknown action: #{action_name}")
    end
  end

  def throw_must_have_actions()
    raise Error.new('You must provide the `actions` parameter to be able to use runActions. ' + LEARN_MORE)
  end

  def get_entities
    req(logger, @access_token, Net::HTTP::Get, "/entities")
  end

  def post_entities(payload)
    payload = payload.map {|k, v| [(k.to_sym rescue k), v]}.to_h.reject{ |k| ![:id, :doc, :values, :lookups].include?(k) }
    validate_payload payload
    req(logger, @access_token, Net::HTTP::Post, "/entities", {}, payload)
  end

  def get_entity(entity_id)
    req(logger, @access_token, Net::HTTP::Get, "/entities/#{URI.encode(entity_id)}")
  end

  def put_entities(entity_id, payload)
    payload = payload.map {|k, v| [(k.to_sym rescue k), v]}.to_h.reject{ |k| ![:id, :doc, :values].include?(k) }
    validate_payload payload
    req(logger, @access_token, Net::HTTP::Put, "/entities/#{URI.encode(entity_id)}", {}, payload)
  end

  def delete_entities(entity_id)
    req(logger, @access_token, Net::HTTP::Delete, "/entities/#{URI.encode(entity_id)}")
  end

  def post_values(entity_id, payload)
    payload = payload.map {|k, v| [(k.to_sym rescue k), v]}.to_h.reject{ |k| ![:value, :expressions, :metadata].include?(k) }
    validate_payload payload
    req(logger, @access_token, Net::HTTP::Post, "/entities/#{URI.encode(entity_id)}/values", {}, payload)
  end

  def delete_values(entity_id, value)
    req(logger, @access_token, Net::HTTP::Delete, "/entities/#{URI.encode(entity_id)}/values/#{URI.encode(value)}")
  end

  def post_expressions(entity_id, value, payload)
    payload = payload.map {|k, v| [(k.to_sym rescue k), v]}.to_h.reject{ |k| ![:expression].include?(k) }
    validate_payload payload
    req(logger,@access_token, Net::HTTP::Post, "/entities/#{URI.encode(entity_id)}/values/#{URI.encode(value)}/expressions", {}, payload)
  end

  def delete_expressions(entity_id, value, expression)
    req(logger,@access_token, Net::HTTP::Delete, "/entities/#{URI.encode(entity_id)}/values/#{URI.encode(value)}/expressions/#{URI.encode(expression)}")
  end

  private

  def validate_payload(payload)
    key_types = {
      id: String,
      doc: String,
      value: String,
      values: Array,
      lookups: Array,
      expression: String,
      expressions: Array,
      metadata: String,
    }
    payload.each do |k, v|
      raise Error.new("#{k.to_s} in request body must be #{key_types[k].to_s} type") unless key_types[k] == v.class
    end
  end

  def __run_actions(session_id, current_request, message, context, i)
    if i <= 0
      raise Error.new('Max steps reached, stopping.')
    end
    json = converse(session_id, message, context)
    if json['type'].nil?
      raise Error.new('Couldn\'t find type in Wit response')
    end
    if current_request != @_sessions[session_id]
      return context
    end

    logger.debug("Context: #{context}")
    logger.debug("Response type: #{json['type']}")

    # backwards-compatibility with API version 20160516
    if json['type'] == 'merge'
      json['type'] = 'action'
      json['action'] = 'merge'
    end

    if json['type'] == 'error'
      raise Error.new('Oops, I don\'t know what to do.')
    end

    if json['type'] == 'stop'
      return context
    end

    request = {
      'session_id' => session_id,
      'context' => context.clone,
      'text' => message,
      'entities' => json['entities']
    }
    if json['type'] == 'msg'
      throw_if_action_missing(:send)
      response = {
        'text' => json['msg'],
        'quickreplies' => json['quickreplies'],
      }
      @actions[:send].call(request, response)
    elsif json['type'] == 'action'
      action = json['action'].to_sym
      throw_if_action_missing(action)
      context = @actions[action].call(request)
      if context.nil?
        logger.warn('missing context - did you forget to return it?')
        context = {}
      end
    else
      raise Error.new("unknown type: #{json['type']}")
    end

    if current_request != @_sessions[session_id]
      return context
    end

    return __run_actions(session_id, current_request, nil, context, i - 1)
  end

  def req(logger, access_token, meth_class, path, params={}, payload={})
    uri = URI(WIT_API_HOST + path)
    uri.query = URI.encode_www_form(params)

    logger.debug("#{meth_class} #{uri}")

    request = meth_class.new(uri)
    request['authorization'] = 'Bearer ' + access_token
    request['accept'] = 'application/vnd.wit.' + WIT_API_VERSION + '+json'
    request.add_field 'Content-Type', 'application/json'
    request.body = payload.to_json

    Net::HTTP.start(uri.host, uri.port, {:use_ssl => uri.scheme == 'https', :read_timeout => @timeout}) do |http|
      rsp = http.request(request)
      if rsp.code.to_i != 200
        raise Error.new("HTTP error code=#{rsp.code}")
      end
      json = JSON.parse(rsp.body)
      if json.is_a?(Hash) and json.has_key?('error')
        raise Error.new("Wit responded with an error: #{json['error']}")
      end
      logger.debug("#{meth_class} #{uri} #{json}")
      json
    end
  end

  def validate_actions(logger, actions)
    [:send].each do |action|
      if !actions.has_key?(action)
        logger.warn "The #{action} action is missing. #{LEARN_MORE}"
      end
    end
    actions.each_pair do |k, v|
      if !k.is_a?(Symbol)
        logger.warn "The '#{k}' action name should be a symbol"
      end
      if !(v.respond_to?(:call) && v.lambda?)
        logger.warn "The '#{k}' action should be a lambda function"
      end
      if k == :send && v.arity != 2
        logger.warn "The \'send\' action should take 2 arguments: request and response. #{LEARN_MORE}"
      end
      if k != :send && v.arity != 1
        logger.warn "The '#{k}' action should take 1 argument: request. #{LEARN_MORE}"
      end
    end
    return actions
  end
end
